### PR TITLE
Separate images for production and development

### PR DESCRIPTION
**Pull request by amercader**
The existing setup works great if you just want to try out CKAN but it's not really suited for development or to customize to your needs.

* Although you can install local extensions with `DEVELOPMENT_MODE=true`, as @gerbyzation pointed out in #3 you can't run the tests because you are missing the dev requirements and other configurations.
* If you want to customize the build for your instance with your own extensions you need to modify the main ckan Dockerfile, which should be stable across releases.

I'm still not 100% happy with the approach but I think this is a step on the right direction. I have separated the images as such:

```
+-----------+                +----------+
|           |                |          |
| ckan-base +---------------->   ckan   | (production)
|           |                |          |
+-----+-----+                +----------+
      |
      |
+-----v----+                 +----------+
|          |                 |          |
| ckan-dev +----------------->   ckan   | (development)
|          |                 |          |
+----------+                 +----------+

```

* `ckan-base`: CKAN with all its dependencies, properly configured and   running on uWSGI (production setup)
* `ckan-dev`: same as the above, plus:

   * Development requirements installed
   * Installs host extensions cloned on the `src` volume
   * Runs on the Paster development server

From these two base images we would build our own customized image tailored to our project, installing any extensions and requirements needed.

To use them users will build their custom images in the `ckan` folder, where there are two Docker files:

* `Dockerfile`: uses `ckan-base`
* `Dockerfile`: uses `ckan-dev`

To use one or the other there are two separate compose files:

* For a production setup: `docker-compose up` (or `docker-compose -f docker-compose.yml up`)
* For a development setup: `docker-compose -f docker-compose.dev.yml up`

There is obviously duplication on the compose files in all the extra services, but I wasn't successful extending the production compose file with a service that used the dev Dockerfile.

I guess we could have a `common.yml` file but I think having this clear separation is worth a bit of duplication.

### CODE REVIEW

**Comment by amercader** [comment]
We'll give a try and iterate over this

**Comment by gerbyzation** [comment]
This looks good, thanks for the great writeup as well. Can learn a thing or two here 🙂

I'll give this a try as well

File: ckan-dev/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+MAINTAINER Open Knowledge International <info@okfn.org>
+
+ENV APP_DIR=/srv/app
+ENV SRC_EXTENSIONS_DIR=/srv/app/src_extensions
+
+# Install packages needed by the dev requirements
+RUN apk add --no-cache libffi-dev
+
+# Install CKAN dev requirements
+RUN pip install --no-binary :all: -r https://raw.githubusercontent.com/ckan/ckan/${GIT_BRANCH}/dev-requirements.txt
```
Comment (gerbyzation) [inline ckan-dev/Dockerfile]: Do we need to pull dev-requirements.txt from github? CKAN is installed in the base image, so we can make a reference to the local one?

File: ckan-base/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+# Create a local user and group to run the app
+RUN addgroup -g 92 -S ckan && \
+    adduser -u 92 -h /srv/app -H -D -S -G ckan ckan
+
+# Create local storage folder
+RUN mkdir -p $CKAN_STORAGE_PATH && \
+    chown -R ckan:ckan $CKAN_STORAGE_PATH
+
+COPY setup ${APP_DIR}
+COPY setup/supervisor.worker.conf /etc/supervisord.d/worker.conf
```
Comment (gerbyzation) [inline ckan-base/Dockerfile]: Should we run the job worker process in the same container? Or is this not possible?

From what I've picked up on advice (I don't have experience with either CKAN jobs or running multiple processes in a container) it's recommended to run a single process. Some of the concerns are around log routing and background processes not picking up on signals.

**Pull request merged**